### ISSUE TITLE
Merged spans disappearing?

### ISSUE THREAD

**Issue by buhrmann** [explosion/spaCy#2890]
Hi, maybe it's too late in the day and I'm not seeing what I'm doing wrong here, but I seem to have disappearing spans when merging:

## How to reproduce the behaviour
```
import en_core_web_sm

nlp = en_core_web_sm.load()

txt = "Today it was my great honor to welcome the 2017 NCAA Football National Champion, Alabama Crimson Tide, to the White House."
doc = nlp(txt)

print("\nNCs before:")
for nc in doc.noun_chunks: print(nc)
    
print("\nEnts before:")
for ent in doc.ents: print(ent)
        
span = list(doc.noun_chunks)[2]
span.merge()

print("\nNCs after:")
for nc in doc.noun_chunks: print(nc)

print("\nEnts after:")
for ent in doc.ents: print(ent)
```

This produces:

```
NCs before:
it
my great honor
the 2017 NCAA Football National Champion
Alabama Crimson Tide
the White House

Ents before:
Today
2017
NCAA Football National Champion
Tide
the White House

NCs after:
it
my great honor
Alabama Crimson Tide
the White House

Ents after:
Today
Tide
the White House
```

I.e. any noun chunk or entity included in the span seems to disappear. The result varies depending on which span is merged. Sometimes both noun chunks and entities disappear, and sometimes just the former.

## Your Environment
* spaCy version      2.0.13         
* Platform           Darwin-17.7.0-x86_64-i386-64bit
* Python version     3.6.6

**Comment by buhrmann** [comment]
Sorry, I also have a follow up question: what would be the correct way to merge both noun_chunks and entities (and possible also other custom matches)? Would I need to collect all spans first into a static list and process them all together, or should it also be possible to first merge all noun_chunks, then entities, and finally matches?

**Comment by buhrmann** [comment]
Never mind... I think the problem was that I assumed the attributes of the merged token would get set automatically. But it seems entity information is lost if not explicitly set in the merge. Also, one has to be more careful to check what exactly one is merging, so that the attributes set on the merged token contain the correct information. For reference, I've found that this seems to work correctly, at least for the test case above:

```python
def merge_noun_phrases(doc):
    """Safely merge noun phrase spans without destroying entity spans."""
    spans = list(doc.noun_chunks)
    for span in spans:
        while len(span) > 1 and span[0].dep_ not in ('advmod', 'amod', 'compound'):
            span = span[1:]
        if len(span) > 1:
            span.merge(tag=span.root.tag_, lemma=span.lemma_, ent_type=span.root.ent_type_)
```

The explicit conversion to lists may not be necessary, I just remembered it being mentioned that changing the spans underneath the generator may not be the safest option...

I think this issue can be closed (or perhaps converted to a documentation issue).
